### PR TITLE
[WIP] ascanrules: add new context analyser for xss in javascript blocks

### PR DESCRIPTION
**Pull request by jokrhub**
Fixes zaproxy/zaproxy#7212

Added a new context analyzer for scanning XSS in JavaScript blocks.

### CODE REVIEW

File: addOns/simpleexample/src/main/resources/org/zaproxy/addon/simpleexample/resources/Messages.properties [hunk 1]
```diff
@@ -1,7 +1,7 @@
 simpleExample.api.action.helloWorld = Logs "hello world called" with debug level.
 simpleExample.desc = A Simple Example Add-on
 simpleExample.error.nofile = File not found : {0}
-simpleExample.panel.msg = <html>This is a simple example of a status panel</html>
+simpleExample.panel.msg = <html>This is a simple example of a status panel. hello chamged 3!!!</html>
```
Comment (kingthorin) [inline addOns/simpleexample/src/main/resources/org/zaproxy/addon/simpleexample/resources/Messages.properties:4]: Revert

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2011 The ZAP Development Team
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:6]: 2022?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import net.htmlparser.jericho.Attribute;
+import net.htmlparser.jericho.Element;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.network.HttpMessage;
+
+
+// parser
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:31]: remove

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+
+    private String scriptBlockCode = null;
+
+    public ScriptContextAnalyser(HttpMessage msg){
+        this.msg = msg;
+        this.htmlPage = msg.getResponseBody().toString();
+        System.err.println("msg:" + this.msg);
+        System.err.println("html:" + this.htmlPage);
+
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:57]: Remove this blank

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContext.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrules.httputils;
+
+import java.lang.annotation.Retention;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.swing.plaf.synth.SynthColorChooserUI;
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContext.java:26]: ?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContext.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrules.httputils;
+
+import java.lang.annotation.Retention;
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContext.java:22]: ?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 7]
```diff
@@ -25,6 +25,9 @@
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+
+import javax.swing.text.TabableView;
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:29]: ?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 8]
```diff
@@ -42,9 +45,21 @@
 import org.zaproxy.addon.commonlib.CommonAlertTag;
 import org.zaproxy.zap.extension.ascanrules.httputils.HtmlContext;
 import org.zaproxy.zap.extension.ascanrules.httputils.HtmlContextAnalyser;
+import org.zaproxy.zap.extension.ascanrules.httputils.ScriptContextAnalyser;
+import org.zaproxy.zap.extension.ascanrules.httputils.ScriptContext;
 import org.zaproxy.zap.model.Vulnerabilities;
 import org.zaproxy.zap.model.Vulnerability;
 
+
+// parser
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:54]: remove

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 9]
```diff
@@ -92,6 +108,58 @@ public class CrossSiteScriptingScanRule extends AbstractAppParamPlugin {
...
+	// 		for (int i=0; i<jsonArr.length(); i++){
+	// 			JSONObject jsonObj = jsonArr.getJSONObject(i);
+
+	// 			if (jsonObj.getString("value").equals(match)){
+	// 				return (jsonObj.getString("type"));
+	// 			}
+	// 		}
+	// }
+
+
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:127]: ?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 10]
```diff
@@ -92,6 +108,58 @@ public class CrossSiteScriptingScanRule extends AbstractAppParamPlugin {
...
+	private static String sendPOST(String params) throws IOException {
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:128]: Please use standard Java camelCase throughout.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 11]
```diff
@@ -92,6 +108,58 @@ public class CrossSiteScriptingScanRule extends AbstractAppParamPlugin {
...
+
+		// For POST only - START
+		con.setDoOutput(true);
+		OutputStream os = con.getOutputStream();
+		os.write(params.getBytes());
+		os.flush();
+		os.close();
+		// For POST only - END
+
+		int responseCode = con.getResponseCode();
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:141]: Use ZAP internals.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 12]
```diff
@@ -92,6 +108,58 @@ public class CrossSiteScriptingScanRule extends AbstractAppParamPlugin {
...
+			}
+			in.close();
+
+			return response.toString();
+
+		
+		} else {
+			return null;
+		}
+	}
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:161]: Use ZAP internals.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 13]
```diff
@@ -240,6 +327,141 @@ private List<HtmlContext> performAttack(
...
+        // additional replace for enc2
+        // attack = attack.replaceFirst("\"", "\\\"");
+        // evidence = attack;
+
+        // System.err.println("printing msg2: " + msg2.getResponseBody());
+
+        // for script
+
+        String responseCode = msg2.getResponseBody().toString();
+        System.err.println("printing response code: " + responseCode);
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:446]: use a logger

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 14]
```diff
@@ -240,6 +327,141 @@ private List<HtmlContext> performAttack(
...
+        } catch (IOException e) {
+            //TODO: handle exception
+            return null;
+        }
+
+        return result;
+
+    }
+
+
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:464]: Use ZAP internals, a logger, and format :grinning:

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 15]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
+        //                 .setParam(param)
+        //                 .setAttack(ctx.getTarget())
+        //                 .setEvidence(ctx.getTarget())
+        //                 .setMessage(ctx.getMsg())
+        //                 .raise();
+        //         return true;
+        //     }
+        // }
+        // return false;
+    // }
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:879]: Remove

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 16]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
-                        .setAttack(ctx.getTarget())
-                        .setEvidence(ctx.getTarget())
-                        .setMessage(ctx.getMsg())
-                        .raise();
-                return true;
-            }
+
+
+        // simple direct attack in single quotes
+        attack = "';alert(1);//";
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:909]: it isn't in single quotes it has a leading single

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 17]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
+        String attack = "default";
+
+
+        // 1. direct attacks - context unaware
+
+        // simple direct attack in double quotes
+        attack = "\";alert(1);//";
+        alertStatus = performAttackForScript(msg, param, attack);
+
+        if (!alertStatus.equals("Empty")){
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:895]: Just don't init, and then use `String.isEmpty()`

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 18]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
+            .setAttack(attack)
+            .setEvidence(context.getTarget())
+            .setMessage(context.getMsg())
+            .raise();
+
+            return true;
+        }
+
+        // escape backslash direct attack in doublequotes
+        attack = "\';alert(1);//";
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:941]: this isn't the same as 924

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 19]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
+        alertStatus = performAttackForScript(msg, param, attack);
+
+        if (!alertStatus.equals("Empty")){
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:944]: as above

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 20]
```diff
@@ -558,33 +780,213 @@ private boolean performCloseTagAttack(HtmlContext context, HttpMessage msg, Stri
...
+            attack = "',x:alert(1),y:'";
+            alertStatus = performAttackForScript(msg, param, attack);
+        }
+        else if (scriptContext.equals("\"")){
+            // double quote payload
+            attack = "\",x:alert(1),y:\"";
+            alertStatus = performAttackForScript(msg, param, attack);
+        }
+
+        if (alertStatus.equals("Empty")){
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:976]: as above

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 21]
```diff
@@ -732,32 +1134,41 @@ public void scan(HttpMessage msg, String param, String value) {
                     return;
                 }
                 hca = new HtmlContextAnalyser(msg2);
-                contexts = hca.getHtmlContexts(value + Constant.getEyeCatcher(), null, 0);
+                contexts = hca.getHtmlContexts(value + "eyeCatcher", null, 0);
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:1137]: Why were these changed?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 22]
```diff
@@ -766,21 +1177,27 @@ public void scan(HttpMessage msg, String param, String value) {
...
                             // Try an img tag
+                            System.err.println(" image tag: " + context);
+                            
                             attackWorked = performImageTagAttack(context, msg, param);
                         }
                     } else {
                         // Last chance - is the payload reflected outside of any
                         // tags
+                        System.err.println(" outside of tags : " + context);
+
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:1193]: Use a logger

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java [hunk 23]
```diff
@@ -766,21 +1177,27 @@ public void scan(HttpMessage msg, String param, String value) {
...
                         attackWorked = performOutsideTagsAttack(context, msg, param);
                     }
                 }
             }
             // Always attack the header if the eyecatcher is reflected in it - this will be
             // different to any alert raised above
-            if (msg2.getResponseHeader().toString().contains(Constant.getEyeCatcher())) {
+            if (msg2.getResponseHeader().toString().contains("eyeCatcher")) {
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:1200]: Why was this changed?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java [hunk 24]
```diff
@@ -86,6 +86,10 @@ public class HtmlContextAnalyser {
     public HtmlContextAnalyser(HttpMessage msg) {
         this.msg = msg;
         this.htmlPage = msg.getResponseBody().toString();
+        
+        System.err.println("msg:" + this.msg);
+        System.err.println("html:" + this.htmlPage);
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java:91]: Use a logger

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    public ScriptContext getScriptContexts(String target){
+        // get script block
+
+        // find context
+        ScriptContext scriptContext = new ScriptContext(this.msg, target);
+        String contextQuote;
+
+        System.err.println("getting script contexts");
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:67]: use a logger

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+
+        String params = "code=" + this.htmlPage + "&target=" + target;
+
+        try {
+            contextQuote = sendPOST(CONTEXT_URL, params); 
+        } catch (Exception e) {
+            return scriptContext;
+        }
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:75]: Use ZAP internals

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+        this.htmlPage = msg.getResponseBody().toString();
+        System.err.println("msg:" + this.msg);
+        System.err.println("html:" + this.htmlPage);
+
+    }
+
+    public ScriptContext getScriptContexts(String target){
+        // get script block
+
+        // find context
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:63]: remove

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java (hunk 26 above)
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:75]: Follow Java naming standards, please

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+			}
+			in.close();
+
+			return response.toString();
+
+		
+		} else {
+			return null;
+		}
+	}
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/ScriptContextAnalyser.java:125]: Use ZAP internals

**Comment by kingthorin** [comment]
Could likely benefit from a spotlessApply as well `./gradlew :aO:ascanrules:spotlessApply`

**Comment by jokrhub** [comment]
Hello @kingthorin. Thanks for the suggestions. I will address all the above issues in the code after the scanner logic for the particular issues has been reviewed. I felt like the scanner logic must be updated to fit more contexts in js. Once I get the feedback on the logic I will rewrite the code according to the standards.

**Comment by kingthorin** [comment]
It's kind of hard to know because the change seem to vast. I don't see any problem with trying to extract context from scripts, that seems like a reasonable way to proceed. It's also unclear what if any current UnitTests are impacted.

Perhaps @psiinon, @thc202, @ricekot have different ideas or interpretation of what's currently here.

**Comment by psiinon** [comment]
For info the external call is a temporary POC - I've been chatting to @jokrhub about this and they understand that the final code must not make external calls or use any technology o ther that java.
The PR has been submitted with this POC code to get feedback on everything else :)

**Comment by kingthorin** [comment]
?

**Comment by thc202** [comment]
It will be open a new PR.

**Comment by kingthorin** [comment]
Okay, great

**Pull request closed without merging**